### PR TITLE
Optimizations

### LINKED ISSUE #99: Creating Carla Map Generator causes crash
When I try to add a "Carla Map Generator" to an empty level, the level editor crashes.

This is the first error message that appears (longer log is available if it helps):
```
[2017.12.20-12.19.47:049][ 29]LogEditor: Attempting to add actor of class 'CarlaMapGenerator' to level at 320.00,-310.00,100.00
Using fixed seed 123456789
[2017.12.20-12.19.47:050][ 29]LogCarla: Generated DCEL with: { 12 vertices, 32 half-edges, 6 faces }
Fatal error: [File:/Storage/wim/Projects/UnrealEngine_4.17/Engine/Source/Runtime/OpenGLDrv/Private/OpenGLVertexDeclaration.cpp] [Line: 91] 
Unknown RHI vertex element type 0
Signal 11 caught.
```

This seems to be an [UE4 4.17.2 issue](https://answers.unrealengine.com/questions/709850/crash-in-4172-with-instanced-static-meshes-unknown.html). But what is the recommendation for this as the recommended UE4 version for Carla is 4.17.2?

### LINKED ISSUE #251: Asynchronous scene captures
Make SceneCaptureCamera capture the images on render thread.

Related to #238.

### LINKED ISSUE #252: Use actors instead of instances in road generator
Improves performance thanks to draw distance and occlusion.

### LINKED ISSUE #253: Add quality level settings
Make quality levels to allow to choose between visual quality and speed.

```ini
[CARLA/QualitySettings]
; Low, Medium, High, or Epic
DefaultLevel=Epic
```

### PR DESCRIPTION
**Pull request by juaxix**
#### Description

Introduce a set of changes that will add optimizations into the CARLA Unreal Engine project. Initially includes Low and Epic quality settings levels.

Changes applied in this branch:

  - Generates pieces for the road
  - Add CARLA settings page in the Unreal Project Settings screen:
  - Draw distance for low quality levels for light, road pieces and other primitive components

###### Fixes

- Fixes #253 Add quality level settings
- Fixes #252 Use actors instead of instances in road generator
- Fixes #251 Screen captures are now being performed inside the render queue
- Closes #99 Creating Carla Map Generator causes crash (no longer an issue)

#### Where has this been tested?

  * **Windows 10** ...
  * **Linux (Ubuntu 16.4):** ...
  * **Unreal Engine 4.18** ...

#### Possible Drawbacks

Check your road generation in the CarlaMapGenerator you may have in your levels to regenerate (using the new method). 
Check your draw distance for meshes and roads.
Check that you are including the new CARLA/QualitySettings -> DefaultLevel in your CarlaSettings.ini. (levels are None, Low or Epic right now)

### CODE REVIEW

File: Unreal/CarlaUE4/Plugins/Carla/Source/Carla/Settings/CarlaSettings.cpp [hunk 1]
```diff
@@ -110,6 +118,19 @@ static void LoadSettingsFromConfig(
...
+
+  // QualitySettings.
+  FString sDefaultLevel;
+  ConfigFile.GetString(S_CARLA_QUALITYSETTINGS, TEXT("DefaultLevel"), sDefaultLevel);
+  if(!Settings.SetQualitySettingsLevel(FQualitySettings::FromString(sDefaultLevel)))
+  {
+	  //error
+  } else
+  {
+	  //apply pre-restart....
```
Comment (nsubiron) [inline Unreal/CarlaUE4/Plugins/Carla/Source/Carla/Settings/CarlaSettings.cpp]: Something to be done here?

File: Unreal/CarlaUE4/Plugins/Carla/Source/Carla/Settings/CarlaSettingsDelegate.cpp [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+
+void UCarlaSettingsDelegate::ApplyQualitySettingsLevelPostRestart() const
+{
+	switch(CarlaSettings->GetQualitySettingsLevel())
+	{
+	  case EQualitySettingsLevel::Low: 
+	  	ApplyLowQualitySettings();
+		  break;
+	  case EQualitySettingsLevel::Medium: 
+      /** @todo */
```
Comment (nsubiron) [inline Unreal/CarlaUE4/Plugins/Carla/Source/Carla/Settings/CarlaSettingsDelegate.cpp]: Maybe we should log an error here warning that it has no effect

File: Unreal/CarlaUE4/Plugins/Carla/Source/Carla/Settings/CarlaSettings.cpp (hunk 1 above)
Comment (juaxix) [reply Unreal/CarlaUE4/Plugins/Carla/Source/Carla/Settings/CarlaSettings.cpp]: Uhm...No, nothing to be done yet, it's planned to be added with operations that would need a reset of the level. But there is no reason to do that right now.

File: Unreal/CarlaUE4/Plugins/Carla/Source/Carla/Settings/CarlaSettingsDelegate.cpp (hunk 2 above)
Comment (juaxix) [reply Unreal/CarlaUE4/Plugins/Carla/Source/Carla/Settings/CarlaSettingsDelegate.cpp]: You are right.

**Comment by juaxix** [comment]
Cars will not change in max culling distance in Epic mode. Maybe it's because of the model itself, but I'll check it, and we can set a maxcullingdistance 0 to be sure in the operations for that quality settings level

**Pull request merged**